posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,926 @@
+subtasks which can be executed in parallel. Getting the final result then
+consists of combining the sub-results of the subtasks. There we can distinguish
+two cases. In the first case, the order in which the sub-results are composed
+doesn't matter, whereas in the second case, it does.
+
+An example of the first case is when the sub-results are numbers and the final
+result is their sum. Obviously, the order in which the numbers are added
+doesn't change the result (to be complete, effects of rounding due to finite
+precision _can_ impact the final result, but that's an interesting but entirely
+different topic). An example where order definitely matters is if we multiply
+a series of matrices: if we split the multiplication into groups, calculate
+the product of the matrices in each group, then the order in which we combine
+the sub-results _does_ matter!
+
+In Scala, we have different tools available to write applications that execute
+code concurrently, and if we execute this code on a multi-CPU or multi-core
+computer, we can achieve parallelism in the execution of this code.
+
+A first tool we have is Akka: Akka allows us to create so-called Actors that
+can execute code. This execution can be triggered by sending an Actor a message.
+In the context of concurrent execution of code, we can imagine a system where
+we create a pool of Actors that execute code concurrently.
+This execution model allows us to implement concurrent execution of sub-tasks
+that fall in the first category mentioned earlier. Unless we take specific
+measures, we have no control over the order in which sub-results are returned.
+Another potential issue with an Actor based implementation is that we don't have
+any flow control. What this means is that if we throw more work at an Actor
+than this Actor can process, work will get piled-up in a buffer (actually,
+the Actor's mailbox). The conclusion is that the Actor model can be applied in a
+straightforward way to problems of the first type, albeit that we need to be
+aware of potential ordering and flow control issues. We'll come back to this
+topic later when we discuss solutions to our sample problem based on
+https://doc.akka.io/docs/akka/current/stream[Akka Streams].
+
+A second tool we have at our disposal is Scala's `Future`. This API allows us
+to tackle both of the aforementioned cases.
+
+Let's see how `Future` can be used to implement a parallelised computation of
+the number π.
+
+== Calculating π
+
+Ok, this topic certainly has been beaten to death, but it's easy to understand,
+so we may just as well use it for demonstration purposes...
+
+First we need an algorithm to calculate π.
+
+I chose the https://en.wikipedia.org/wiki/Bailey–Borwein–Plouffe_formula[_Bailey--Borwein--Plouffe formula_] _(BBP)_
+discovered in 1995 by _Simon Plouffe_ :
+
+image::2020-03-10-calculating-pi-concurrently/Pi-formula.png[BBP digit extraction algorithm][{half-width}]
```

Comment:
Is that a sideways 8? ;)

For those unfamiliar with summations, perhaps it could be nice to explain this algorithm in english or pseudocode as well?

Reply:
Good suggestion